### PR TITLE
TI Project requirements

### PR DESCRIPTION
**Pull request by SergejMuhic**
1. created Project root with requirements

EDIT: Please review the entire file since there was a baseline committed beforehand.

### CODE REVIEW

File: IFC4x3/ModelViews/TI-Bridge/DocModelView.xml [hunk 1]
```diff
@@ -86,6 +86,55 @@
...
+					<Documentation>The name or code identifying the project within the owning agency.</Documentation>
+					<Definition xsi:type="DocTemplateDefinition" Name="Name" UniqueId="69fa0453-2978-418a-b021-0251fef8a5f4" Type="IfcProject" />
+				</DocTemplateUsage>
+				<DocTemplateUsage Name="Length Unit" UniqueId="57cf383a-8757-4a12-8166-e6e3f9b061c0" Operator="or">
+					<Documentation>Default unit for length. For imperial, inches are recommended. SI Meters are the default if not provided.</Documentation>
+					<Definition xsi:type="DocTemplateDefinition" xsi:nil="true" href="Project_Units_1CogmCYFXC7P7z8XJGvHF4" />
+					<Items>
+						<DocTemplateItem UniqueId="dd3cdeb3-b37d-4910-87a7-6cf081e954ac" RuleParameters="HasUnits=IfcUnitAssignment;SIUnitType=LENGTHUNIT;SIUnitName=METRE;" />
+						<DocTemplateItem UniqueId="13d50766-f8b6-4166-b634-ab9ebae77c6b" RuleParameters="HasUnits=IfcUnitAssignment;ConversionUnitType=LENGTHUNIT;ConversionUnitName=inch;" />
+					</Items>
```
Comment (pjanck) [inline IFC4x3/ModelViews/TI-Bridge/DocModelView.xml]: I believe that this syntax is 1.0. We might consider updating the syntax to be in line with version 1.1 or 1.2.

File: IFC4x3/ModelViews/TI-Bridge/DocModelView.xml [hunk 2]
```diff
@@ -86,6 +86,55 @@
...
+						<DocTemplateItem UniqueId="23d6a36d-e096-4d88-9d2b-2c07d9ad7f38" RuleParameters="HasUnits=IfcUnitAssignment;ConversionUnitType=PLANEANGLEUNIT;ConversionUnitName=degree;" />
+						<DocTemplateItem UniqueId="638ecb0b-2dc0-4445-a898-0e58a97305f6" RuleParameters="HasUnits=IfcUnitAssignment;SIUnitType=PLANEANGLEUNIT;SIUnitName=RADIAN;" />
+					</Items>
+				</DocTemplateUsage>
+				<DocTemplateUsage Name="Mass Unit" UniqueId="88f2eb0b-e99d-4a58-909c-8621c7c6faf6">
+					<Documentation>Default unit for mass. For imperial, pounds are recommended. SI Kilograms are the default if not provided.</Documentation>
+					<Definition xsi:type="DocTemplateDefinition" xsi:nil="true" href="Project_Units_1CogmCYFXC7P7z8XJGvHF4" />
+					<Items>
+						<DocTemplateItem UniqueId="986637a8-b557-46e2-826d-f2ba63ee6ac7" RuleParameters="HasUnits=IfcUnitAssignment;SIUnitType=MASSUNIT;SIUnitName=GRAM;" />
+					</Items>
```
Comment (pjanck) [inline IFC4x3/ModelViews/TI-Bridge/DocModelView.xml]: Let's add (similarly to length units) a conversion based unit for pounds.

File: IFC4x3/ModelViews/TI-Bridge/DocModelView.xml [hunk 3]
```diff
@@ -86,6 +86,55 @@
...
+				</DocTemplateUsage>
+				<DocTemplateUsage Name="Temperature Unit" UniqueId="5f203dc7-daa5-4269-a0ca-ab493c7961dd" Operator="or">
+					<Documentation>Default unit for temperature. For imperial, fahrenheit is recommended. SI Celcius is the default if not provided.</Documentation>
+					<Definition xsi:type="DocTemplateDefinition" xsi:nil="true" href="Project_Units_1CogmCYFXC7P7z8XJGvHF4" />
+					<Items>
+						<DocTemplateItem UniqueId="8ef2e7b3-e250-4aa4-9473-047c29cff927" RuleParameters="HasUnits=IfcUnitAssignment;SIUnitType=THERMODYNAMICTEMPERATUREUNIT;SIUnitName=DEGREE_CELSIUS;" />
+						<DocTemplateItem UniqueId="6e581b24-9e39-4dff-8a48-16364fbbbf1f" RuleParameters="HasUnits=IfcUnitAssignment;SIUnitType=THERMODYNAMICTEMPERATUREUNIT;" />
+					</Items>
+				</DocTemplateUsage>
```
Comment (pjanck) [inline IFC4x3/ModelViews/TI-Bridge/DocModelView.xml]: At imperial check (I assume that would be line 125):
- Fahrenheit is missing from checking parameters
- ConversionUnitType RuleID should be used, and not SIUnitType

File: IFC4x3/ModelViews/TI-Bridge/DocModelView.xml [hunk 4]
```diff
@@ -86,6 +86,55 @@
...
+				<DocTemplateUsage Name="Site" UniqueId="cd0dc4fc-ae19-46f7-a9ee-c2b84b8bc0c1">
+					<Documentation>Site within project describing geospatial location and project boundaries.</Documentation>
+					<Definition xsi:type="DocTemplateDefinition" xsi:nil="true" href="Project_Global_Positioning_3J9e0D0k14GeJdkFjFtWG0" />
```
Comment (pjanck) [inline IFC4x3/ModelViews/TI-Bridge/DocModelView.xml:118]: I am confused about `Site` mention, however the `href` points to `Project_Global_Positioning..`: the latter uses `IfcGeometricRepresentationContext` and not `IfcSite`. Is this a mistake or just a similar naming without intent to imply any connection?

**Review by pjanck — CHANGES_REQUESTED** [review]
I've only read through the xml code and did not load this in any UI.

File: IFC4x3/ModelViews/TI-Bridge/DocModelView.xml (hunk 4 above)
Comment (SergejMuhic) [reply IFC4x3/ModelViews/TI-Bridge/DocModelView.xml:118]: Not sure I understand the question. It is a judgement from the description of the requirement. Names are just identifiers, the requirement describes the test. We can argue about what to do about it but from my point of view "Site within project describing geospatial location and project boundaries." requires geospatial location first and foremost. Project boundaries I honestly do not understand.

**Review by pjanck — COMMENTED** [review]
Reviewed the last two commits only. I still need to go through the whole content as indicated in the PR's description.

File: IFC4x3/ModelViews/TI-Bridge/DocModelView.xml [hunk 5]
```diff
@@ -71,7 +71,57 @@
...
+					<Items>
+						<DocTemplateItem UniqueId="23d6a36d-e096-4d88-9d2b-2c07d9ad7f38" RuleParameters="HasUnits=IfcUnitAssignment;ConversionUnitType=PLANEANGLEUNIT;ConversionUnitName=degree;" />
+						<DocTemplateItem UniqueId="638ecb0b-2dc0-4445-a898-0e58a97305f6" RuleParameters="HasUnits=IfcUnitAssignment;SIUnitType=PLANEANGLEUNIT;SIUnitName=RADIAN;" />
+					</Items>
+				</DocTemplateUsage>
+				<DocTemplateUsage Name="Mass Unit" UniqueId="88f2eb0b-e99d-4a58-909c-8621c7c6faf6">
+					<Documentation>Default unit for mass. For imperial, pounds are recommended. SI Kilograms are the default if not provided.</Documentation>
+					<Definition xsi:type="DocTemplateDefinition" xsi:nil="true" href="Project_Units_1CogmCYFXC7P7z8XJGvHF4" />
+					<Items>
+						<DocTemplateItem UniqueId="986637a8-b557-46e2-826d-f2ba63ee6ac7" RuleParameters="HasUnits=IfcUnitAssignment;SIUnitType=MASSUNIT;SIUnitName=GRAM;" />
```
Comment (pjanck) [inline IFC4x3/ModelViews/TI-Bridge/DocModelView.xml]: In order to conform to "SI Kilograms are the default", one would need to include a check for `SIPrefix[VALUE]='KILO'`.

**Review by pjanck — CHANGES_REQUESTED** [review]
Using mvdXML Grammar to define rules, not the encoding in the reviewed file:

File: IFC4x3/ModelViews/TI-Bridge/DocModelView.xml [hunk 6]
```diff
@@ -71,7 +71,57 @@
...
+			<Concepts>
+				<DocTemplateUsage Name="Name" UniqueId="f10112e6-72f5-4d58-ad80-b8456e33b3f1">
+					<Documentation>The name or code identifying the project within the owning agency.</Documentation>
+					<Definition xsi:type="DocTemplateDefinition" Name="Name" UniqueId="69fa0453-2978-418a-b021-0251fef8a5f4" Type="IfcProject" />
+				</DocTemplateUsage>
+				<DocTemplateUsage Name="Length Unit" UniqueId="57cf383a-8757-4a12-8166-e6e3f9b061c0" Operator="or">
+					<Documentation>Default unit for length. For imperial, inches are recommended. SI Meters are the default if not provided.</Documentation>
+					<Definition xsi:type="DocTemplateDefinition" xsi:nil="true" href="Project_Units_1CogmCYFXC7P7z8XJGvHF4" />
+					<Items>
+						<DocTemplateItem UniqueId="dd3cdeb3-b37d-4910-87a7-6cf081e954ac" RuleParameters="HasUnits=IfcUnitAssignment;SIUnitType=LENGTHUNIT;SIUnitName=METRE;" />
```
Comment (pjanck) [inline IFC4x3/ModelViews/TI-Bridge/DocModelView.xml]: Similarly to kilograms, one would need to check that `SIPrefix[EXISTS]=FALSE` for the case of "SI Meters are the default". Compare:

    #1 = IFCSIUNIT(*, .LENGTHUNIT., $, .METRE.);
    #2 = IFCSIUNIT(*, .LENGTHUNIT., .MILLI., .METRE.);

Both lines would pass the current rule.

**Review by pjanck — CHANGES_REQUESTED** [review]
Goes with previous review (forgot to submit ...).

File: IFC4x3/ModelViews/TI-Bridge/DocModelView.xml (hunk 5 above)
Comment (SergejMuhic) [reply IFC4x3/ModelViews/TI-Bridge/DocModelView.xml]: This should be fine now.

File: IFC4x3/ModelViews/TI-Bridge/DocModelView.xml (hunk 6 above)
Comment (SergejMuhic) [reply IFC4x3/ModelViews/TI-Bridge/DocModelView.xml]: Have a proposal committed on how to deal with it in the spec. I'd have to implement serialization. Let me know what you think.

**Review by pjanck — APPROVED** [review]
LFTM

Very neat solution, adding the `SIPrefix` RuleID!

**Review by pjanck — APPROVED** [review]

**Review by larswik — APPROVED** [review]

**Pull request merged**